src/pymatching/sparse_blossom/driver/user_graph.test.cc

```diff
@@ -166,3 +181,144 @@ TEST(UserGraph, DecodeUserGraphDetectionEventOnBoundaryNode) {
+TEST(IterDemInstructionsTest, ErrorWithOneObservable) {
+    stim::DetectorErrorModel dem("error(0.125) D1 D2 L0");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 1);
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.125, 1, 2, {0}}));
+}
+
+// Test an error that flips multiple logical observables.
+TEST(IterDemInstructionsTest, ErrorWithMultipleObservables) {
+    stim::DetectorErrorModel dem("error(0.3) D3 D4 L1 L3");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 1);
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.3, 3, 4, {1, 3}}));
+}
+
+// Test an error with probability 0. It should be ignored.
+TEST(IterDemInstructionsTest, ZeroProbabilityError) {
+    stim::DetectorErrorModel dem("error(0.0) D0 D1");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_TRUE(handler.handled_errors.empty());
+}
+
+// Test an error involving more than two detectors (a hyperedge). This should be ignored.
+TEST(IterDemInstructionsTest, ThreeDetectorErrorIsIgnored) {
+    stim::DetectorErrorModel dem("error(0.1) D0 D1 D2");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_TRUE(handler.handled_errors.empty());
+}
+
+// Test a decomposed error instruction. The handler should be called for each component.
+TEST(IterDemInstructionsTest, DecomposedError) {
+    stim::DetectorErrorModel dem("error(0.1) D0 D1 ^ D2 D3 L0 ^ D4");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 3);
+
+    EXPECT_EQ(handler.handled_errors[0], (HandledError{0.1, 0, 1, {}}));
+    EXPECT_EQ(handler.handled_errors[1], (HandledError{0.1, 2, 3, {0}}));
+    EXPECT_EQ(handler.handled_errors[2], (HandledError{0.1, 4, SIZE_MAX, {}}));
+}
+
+// Test a decomposed error where one of the components is a hyperedge and should be ignored.
+TEST(IterDemInstructionsTest, DecomposedErrorWithIgnoredComponent) {
+    stim::DetectorErrorModel dem("error(0.15) D0 D1 ^ D2 D3 D4 ^ D5 D6 L2");
+    TestHandler handler;
+    pm::iter_dem_instructions_include_correlations(dem, handler);
+    ASSERT_EQ(handler.handled_errors.size(), 2);
+
+    std::vector<HandledError> expected = {{0.15, 0, 1, {}}, {0.15, 5, 6, {2}}};
+    EXPECT_EQ(handler.handled_errors, expected);
+}
+
+// Test a complex DEM with multiple instruction types and edge cases combined.
+TEST(IterDemInstructionsTest, CombinedComplexDem) {
+    stim::DetectorErrorModel dem(R"DEM(
+        error(0.1) D0            # Instruction 1: Simple
+        error(0.2) D1 D2 L0      # Instruction 2: Two detectors, one observable
+        error(0.3) D3 D4 D5 ^ D6 # Instruction 3: Hyperedge ignored, second component handled
```

Comment:
Could replace this with a single hyperedge:

`error(0.3) D3 D4 D5`
 
since I think it is fine to ignore this rather than throwing an exception.